Simple Log4j tests with in-memory log appender

kshepherd opened the pull request:
Related to https://github.com/DSpace/DSpace/issues/11046

This integration test creates and attaches an in-memory log appender, writes an INFO line, and tests for expected output.

It is a controller integration test rather than a simple abstract test, not to test any specific REST webapp functionality but to force it to be run in that environment, as earlier log4j bugs appeared to affect logging from Tomcat but not the CLI (plain kernel).

- [ ] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [ ] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [ ] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [ ] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [ ] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [ ] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [ ] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

kshepherd committed:
Add simple log4j WebappLoggingIT test

tdonohue commented:
@kshepherd : For some strange reason none of the automated tests are running for this PR (they seem to have failed to trigger, not sure why).  

But, I've noted a compilation failure above that I found locally.  Perhaps once that's fixed, it'll trigger automated tests to run here properly.

kshepherd commented:
@tdonohue hm, i'll look at that - i was already confused as my local env was not running the test when i used the usual `mvn install -DskipIntegrationTests=false -Dit.test=org.dspace.rest.test.WebappLoggingIT` (my maven and jdk often gets confused with me switching versions so often so i was just blaming that)

kshepherd commented:
thanks, changed that to an `@After` anyway since if more tests are added, we do want that cleanup after every test. fixed other issues and manually ran checkstyle

kshepherd commented:
checks now pass but unfortunately, if i switch my log4j version back to `2.25.0`, recompile and test again, it still passes. so whatever i'm doing with that appender is not triggering the condition seen in our file log appenders / prod log4j config

tdonohue commented:
@kshepherd : As noted in other comments, I have an alternative or perhaps complimentary PR in #11077 which can detect the issues with log4j 2.25.0.   That said, I'm not against also merging this PR as a safety measure -- it's possible this PR will detect other logging issues, even though it doesn't detect the flaw encountered with log4j 2.25.0.

tdonohue reviewed (approved):
👍 Thanks @kshepherd .  Even though this didn't catch the log4j failure we encountered with 2.25.0, I still feel this test is worth adding "just in case".  It is possible this test could be beneficial in catching or diagnosing future logging issues.
